File: python/evm_state/ingest.py
"""Bind one finalized native sink to one immutable package, filter and database.

Cursor loss is explicit recovery work: never silently start an old account list
over a database belonging to another run. Keep this directory on durable storage.
"""
import hashlib
import json
import os
from pathlib import Path
import subprocess
import urllib.parse
import uuid

from .checkpoint import canonical_accounts, connect_like
from .files import atomic_json, atomic_write, exclusive_lock
from .proof import VerificationError
from .ch import identifier
from .cursor import load_progress, observe, save_progress
from .capacity import check as capacity_check
from . import host

MODULE = "map_block_state"


def _native(command, dsn):
    env = dict(os.environ, SUBSTREAMS_SINK_DSN=dsn)
    result = subprocess.run(command, env=env, capture_output=True, text=True, timeout=120)
    if result.returncode:
        # DSNs can appear in third-party errors. Do not echo credentials.
        raise RuntimeError("native Substreams command failed: " + result.stderr[-4000:].replace(dsn, "[database]"))
    return result.stdout


def _identity(client, spkg, endpoint, accounts, start_block, directory, dsn, checkpoint_database=None):
    if isinstance(start_block, bool) or not isinstance(start_block, int) or start_block < 0:
        raise ValueError("start block must be an absolute nonnegative integer")
    parsed = urllib.parse.urlsplit(dsn)
    http = urllib.parse.urlsplit(client.url)
    normalize_host = lambda value: "127.0.0.1" if value == "localhost" else value
    if parsed.scheme != "clickhouse" or parsed.path != "/" + client.database or normalize_host(parsed.hostname) != normalize_host(http.hostname):
        raise ValueError("native DSN and HTTP client must name the same ClickHouse host and database")
    if http.username or http.password or http.query or http.fragment:
        raise ValueError("use CH_USER/CH_PASSWORD for HTTP credentials")
    selected = canonical_accounts(accounts)
    package = Path(spkg).read_bytes()
    info = json.loads(_native(["substreams", "info", str(spkg), "--json", "-p", MODULE + "=" + ",".join(selected)], dsn))
    if Path(spkg).read_bytes() != package:
        raise VerificationError("package changed during preparation; retry after the build finishes")
    module = next((v for v in info["modules"] if v["name"] == MODULE), {})
    if info.get("network") != "bsc" or module.get("output_type") != "proto:evm.state.v1.BlockState":
        raise VerificationError("expected the BSC native block-state package")
    local_host = host.machine_id()
    previous = directory / "run.json"
    if previous.exists():
        previous_identity = json.loads(previous.read_text())["identity"]
        if host.matches(previous_identity, directory):
            local_host = previous_identity["host"]
    return {"format_version": 3, "database": client.database, "http_url": client.url,
        "checkpoint_database": identifier(checkpoint_database or client.database),
        "state_directory": str(directory), "host": local_host,
        "native_target": f"{parsed.hostname}:{parsed.port or 9000}/{client.database}",
        "endpoint": endpoint, "accounts": selected, "start_block": start_block,
        "module": MODULE, "module_hash": module["hash"], "package_sha256": hashlib.sha256(package).hexdigest(),
        "network": "bsc", "schema_version": 1, "final_blocks_only": True}, package


def _prepare(client, spkg, endpoint, accounts, start_block, directory, dsn, checkpoint_database=None):
    identity, package = _identity(client, spkg, endpoint, accounts, start_block, directory, dsn, checkpoint_database)
    record_path = directory / "run.json"
    admin = connect_like(client, "default")
    exists = bool(int(admin.one("SELECT count() AS n FROM system.databases WHERE name={db:String}",
                                {"db": client.database})["n"]))
    if record_path.exists():
        record = json.loads(record_path.read_text())
        if record.get("phase") not in {"preparing", "prepared"}:
            raise VerificationError("invalid native run phase")
        if record["identity"] != identity:
            raise VerificationError("run identity changed; use a new isolated database and state directory")
        if record["phase"] == "prepared" and not exists:
            raise VerificationError("run database is missing; restore matching database and cursor metadata")
    else:
        if exists and int(client.one("SELECT count() AS n FROM system.tables WHERE database={db:String}",
                                      {"db": client.database})["n"]):
            raise VerificationError("database already contains tables without this run's local identity")
        if any(directory.iterdir()):
            unexpected = [p.name for p in directory.iterdir() if p.name != "run.lock"]
            if unexpected:
                raise VerificationError("state directory is not empty and has no run identity")
        record = {"run_id": uuid.uuid4().hex, "identity": identity, "phase": "preparing"}
        atomic_json(record_path, record)
    frozen = directory / "package.spkg"
    if frozen.exists():
        if hashlib.sha256(frozen.read_bytes()).hexdigest() != identity["package_sha256"]:
            raise VerificationError("frozen run package was changed")
    elif record["phase"] == "prepared":
        raise VerificationError("frozen run package is missing")
    else:
        atomic_write(frozen, package)
    admin.execute(f"CREATE DATABASE IF NOT EXISTS {client.database}")
    db_uuid = client.one("SELECT toString(uuid) AS uuid FROM system.databases WHERE name={db:String}",
                         {"db": client.database})["uuid"]
    if record.get("database_uuid", db_uuid) != db_uuid:
        raise VerificationError("database was replaced; restore matching database and run metadata")
    owner_exists = int(client.one("SELECT count() AS n FROM system.tables WHERE database={db:String} AND name='_evm_state_run'",
                                  {"db": client.database})["n"])
    if not owner_exists:
        if record["phase"] == "prepared":
            raise VerificationError("database run identity is missing")
        # No IF NOT EXISTS: two different state directories cannot both claim a DB.
        client.execute("CREATE TABLE _evm_state_run (run_id String, identity String) ENGINE=MergeTree ORDER BY run_id "
                       "SETTINGS fsync_after_insert=1, fsync_part_directory=1")
        client.insert("_evm_state_run", [{"run_id": record["run_id"], "identity": json.dumps(identity, sort_keys=True)}])
    owner = client.one("SELECT run_id,identity FROM _evm_state_run")
    if owner["run_id"] != record["run_id"] or json.loads(owner["identity"]) != identity:
        raise VerificationError("database belongs to a different native run")
    metadata = directory / "meta" / f"{client.database}_schema_hash.txt"
    if record["phase"] == "preparing":
        metadata.parent.mkdir(exist_ok=True)
        _native(["substreams", "sink", "clickhouse", "setup", str(frozen), MODULE,
            "--bytes-encoding", "0xhex", "--sink-info-folder", str(metadata.parent)], dsn)
        # One-row block envelopes are atomic; also require durable acknowledged parts.
        for table in ["state_blocks", "_blocks_"]:
            client.execute(f"ALTER TABLE {table} MODIFY SETTING fsync_after_insert=1, fsync_part_directory=1")
        with metadata.open("rb") as handle:
            os.fsync(handle.fileno())
        record.update(phase="prepared", database_uuid=db_uuid, schema_hash=metadata.read_text().strip())
        atomic_json(record_path, record, overwrite=True)
    if not metadata.is_file() or metadata.read_text().strip() != record["schema_hash"]:
        raise VerificationError("native schema metadata is missing or changed")
    return record


def prepare(client, spkg, endpoint, accounts, start_block, directory, dsn, checkpoint_database=None):
    directory = Path(directory).resolve()
    capacity_check(client, [directory], "native-prepare")
    with exclusive_lock(directory / "run.lock"):
        return _prepare(client, spkg, endpoint, accounts, start_block, directory, dsn, checkpoint_database)


def recover_cursor(client, spkg, endpoint, accounts, start_block, directory, dsn, checkpoint_database=None):
    directory = Path(directory).resolve()
    with exclusive_lock(directory / "run.lock"):
        run = _prepare(client, spkg, endpoint, accounts, start_block, directory, dsn, checkpoint_database)
        progress = load_progress(client, run, directory)
        cursor = directory / "cursor.txt"
        # Preserve the damaged cursor as evidence without treating it as trusted
        # progress. Never overwrite a directory or silently adopt an older run.
        if cursor.is_file():
            atomic_write(directory / ("cursor-before-recovery-" + uuid.uuid4().hex + ".txt"), cursor.read_bytes())
        atomic_write(cursor, progress["cursor"].encode(), overwrite=True)
        return {"run_id": run["run_id"], "recovered": True, "position": progress["position"]}


def validate_parallel_workers(value):
    if value is not None and (isinstance(value, bool) or not isinstance(value, int) or value < 1):
        raise ValueError("parallel workers must be a positive integer or omitted")


def ingest(client, spkg, endpoint, accounts, start_block, directory, dsn, stop_block=None, max_retries=3,
           checkpoint_database=None, decode_batch_size=1, spool_max_idle_ms=100, prometheus_addr=None,
           parallel_workers=None):
    validate_parallel_workers(parallel_workers)
    directory = Path(directory).resolve()
    if stop_block is not None and stop_block <= start_block:
        raise ValueError("stop block must be greater than start block (exclusive)")
    if isinstance(decode_batch_size, bool) or not isinstance(decode_batch_size, int) or decode_batch_size < 1:
        raise ValueError("decode batch size must be a positive integer")
    if isinstance(spool_max_idle_ms, bool) or not isinstance(spool_max_idle_ms, int) or spool_max_idle_ms < 1:
        raise ValueError("spool maximum idle milliseconds must be a positive integer")
    with exclusive_lock(directory / "run.lock") as run_lock:
        capacity_check(client, [directory], "native-start")
        record = _prepare(client, spkg, endpoint, accounts, start_block, directory, dsn, checkpoint_database)
        cursor = directory / "cursor.txt"
        blocks = int(client.one("SELECT count() AS n FROM state_blocks FINAL")["n"])
        if blocks and (not cursor.is_file() or not cursor.read_text().strip()):
            raise VerificationError("native cursor is missing or empty for existing data; explicit recovery is required")
        if not blocks and cursor.exists():
            raise VerificationError("cursor exists without its block data; restore matching database and metadata")
        report = {"run_id": record["run_id"], "source": {"database": client.database,
            "accounts": record["identity"]["accounts"], "start_block": start_block,
            "module_hash": record["identity"]["module_hash"], "final_blocks_only": True}}
        if blocks:
            progress = save_progress(client, record, directory, cursor.read_text().strip())
            if stop_block is not None and progress["position"]["block"]["number"] >= stop_block - 1:
                if progress["position"]["block"]["number"] != stop_block - 1:
                    raise VerificationError("requested stop precedes the already ingested cursor")
                return {**report, "already_complete": True, "position": progress["position"]}
        elif (directory / "durable_progress.json").exists():
            raise VerificationError("durable progress exists without its block data")
        command = ["substreams", "sink", "clickhouse", str(directory / "package.spkg"), MODULE,
            "-e", endpoint, "-p", MODULE + "=" + ",".join(record["identity"]["accounts"]), "-s", str(start_block),
            "--final-blocks-only", "--bytes-encoding", "0xhex", "--sink-info-folder", str(directory / "meta"),
            "--cursor-file-path", str(cursor), "--spool-dir", str(directory / "spool"), "--spool-max-size", "1GiB",
            "--spool-max-idle", f"{spool_max_idle_ms}ms", "--max-retries", str(max_retries)]
        # The pinned sink's direct-insert transition requires STEP_NEW. A
        # finalized-only follow stays STEP_NEW_IRREVERSIBLE even at the head,
        # so it needs a one-block decode batch and a short spool idle timeout.
        # Otherwise the default 32-block batch adds ~15 seconds on BSC, and
        # one-block batches alone can keep the spool open until its size target.
        command.extend(["--decode-batch-size", str(decode_batch_size)])
        if parallel_workers is not None:
            command.extend(["--header", f"X-Substreams-Parallel-Workers:{parallel_workers}"])
        if prometheus_addr is not None:
            command.extend(["--prometheus-addr", prometheus_addr])
        if stop_block is not None:
            command.extend(["-t", str(stop_block)])
        # If this wrapper is SIGKILLed, its native child may remain alive. Keep
        # the same flock open in that child so another wrapper cannot take over
        # while the orphan still writes. Normal cleanup waits for the child
        # before releasing the lock. Supported platforms are POSIX (Linux/macOS).
        process = subprocess.Popen(command, env=dict(os.environ, SUBSTREAMS_SINK_DSN=dsn),
                                   pass_fds=(run_lock.fileno(),))
        try:
            while True:
                try:
                    result = process.wait(timeout=1)
                    break
                except subprocess.TimeoutExpired:
                    observe(client, record, directory)
                    capacity_check(client, [directory], "native-progress")
            if result:
                raise RuntimeError(f"native sink exited with status {result}; retain its cursor and spool for recovery")
        finally:
            if process.poll() is None:
                process.terminate()
                try:
                    process.wait(timeout=15)
                except subprocess.TimeoutExpired:
                    process.kill(); process.wait()
            # SIGTERM may drain a pending spool and write a newer valid cursor.
            # Back it up even when this run exits through a capacity/error path.
            progress = observe(client, record, directory)
        if progress is None:
            raise VerificationError("native sink completed without a valid cursor; recover from durable progress")
        if stop_block is not None and progress["position"]["block"]["number"] != stop_block - 1:
            raise VerificationError("native sink stopped before its requested final block")
        with cursor.open("rb") as handle:
            os.fsync(handle.fileno())
        atomic_write(directory / "last_completed_cursor.txt", progress["cursor"].encode(), overwrite=True)
        return {**report, "position": progress["position"]}

File: crates/evm-state/src/ingest.rs
//! Guarded native finalized ingestion with immutable source identity and recovery.
use crate::{
    capacity,
    ch::{identifier, params, uint, ClickHouse},
    checkpoint::canonical_accounts,
    control::new_id,
    cursor,
    files::{atomic_json, atomic_write, file_lock, resolve, spaced_json},
    host, process,
    proof::string,
};
use anyhow::{ensure, Context, Result};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::{
    fs,
    path::{Path, PathBuf},
    process::Command,
    sync::{
        atomic::{AtomicBool, Ordering},
        Arc,
    },
    time::{Duration, Instant},
};
use wait_timeout::ChildExt;

pub const MODULE: &str = "map_block_state";

#[derive(Clone)]
pub struct NativeOptions {
    pub package: PathBuf,
    pub endpoint: String,
    pub accounts: Value,
    pub start_block: u64,
    pub state_dir: PathBuf,
    pub dsn: String,
    pub checkpoint_database: Option<String>,
}

#[derive(Clone)]
pub struct IngestOptions {
    pub stop_block: Option<u64>,
    pub max_retries: i64,
    pub decode_batch_size: u32,
    pub spool_max_idle_ms: u64,
    pub prometheus_addr: Option<String>,
    pub parallel_workers: Option<u32>,
}
impl Default for IngestOptions {
    fn default() -> Self {
        Self {
            stop_block: None,
            max_retries: 3,
            decode_batch_size: 1,
            spool_max_idle_ms: 100,
            prometheus_addr: None,
            parallel_workers: None,
        }
    }
}
impl IngestOptions {
    pub fn validate(&self, start: u64) -> Result<()> {
        ensure!(
            self.parallel_workers.is_none_or(|v| v > 0),
            "parallel workers must be a positive integer or omitted"
        );
        ensure!(
            self.stop_block.is_none_or(|stop| stop > start),
            "stop block must be greater than start block (exclusive)"
        );
        ensure!(
            self.decode_batch_size > 0,
            "decode batch size must be a positive integer"
        );
        ensure!(
            self.spool_max_idle_ms > 0,
            "spool maximum idle milliseconds must be a positive integer"
        );
        Ok(())
    }
}

fn native(arguments: &[String], dsn: &str) -> Result<String> {
    let output = process::capture(
        Command::new("substreams")
            .args(arguments)
            .env("SUBSTREAMS_SINK_DSN", dsn),
        Duration::from_secs(120),
    )?
    .context("native Substreams command timed out")?;
    ensure!(
        output.status.success(),
        "native Substreams command failed (status {}); inspect the retained native state",
        process::exit_code(output.status)
    );
    String::from_utf8(output.stdout).context("invalid native Substreams response")
}

pub fn validate_target(client: &ClickHouse, dsn: &str) -> Result<reqwest::Url> {
    let parsed =
        reqwest::Url::parse(dsn).map_err(|_| anyhow::anyhow!("invalid native ClickHouse DSN"))?;
    let http = reqwest::Url::parse(&client.url)
        .map_err(|_| anyhow::anyhow!("invalid ClickHouse HTTP URL"))?;
    let normalize = |host: Option<&str>| {
        host.map(|value| {
            if value == "localhost" {
                "127.0.0.1".to_owned()
            } else {
                value.to_owned()
            }
        })
    };
    ensure!(
        parsed.scheme() == "clickhouse"
            && parsed.path() == format!("/{}", client.database)
            && normalize(parsed.host_str()) == normalize(http.host_str()),
        "native DSN and HTTP client must name the same ClickHouse host and database"
    );
    ensure!(
        http.username().is_empty()
            && http.password().is_none()
            && http.query().is_none()
            && http.fragment().is_none(),
        "use CH_USER/CH_PASSWORD for HTTP credentials"
    );
    Ok(parsed)
}

fn identity(
    client: &ClickHouse,
    options: &NativeOptions,
    directory: &Path,
) -> Result<(Value, Vec<u8>)> {
    let parsed = validate_target(client, &options.dsn)?;
    let selected = canonical_accounts(&options.accounts)?;
    let package = fs::read(&options.package)?;
    let info: Value = serde_json::from_str(&native(
        &[
            "info".into(),
            options
                .package
                .to_str()
                .context("invalid package path")?
                .into(),
            "--json".into(),
            "-p".into(),
            format!("{MODULE}={}", selected.join(",")),
        ],
        &options.dsn,
    )?)?;
    ensure!(
        fs::read(&options.package)? == package,
        "package changed during preparation; retry after the build finishes"
    );
    let module = info["modules"]
        .as_array()
        .context("missing native module metadata")?
        .iter()
        .find(|v| v["name"] == MODULE)
        .context("native block-state module is missing")?;
    ensure!(
        info["network"] == "bsc" && module["output_type"] == "proto:evm.state.v1.BlockState",
        "expected the BSC native block-state package"
    );
    let mut local_host = host::machine_id()?;
    if directory.join("run.json").try_exists()? {
        let previous: Value = serde_json::from_slice(&fs::read(directory.join("run.json"))?)?;
        if host::matches(&previous["identity"], directory)? {
            local_host = string(&previous["identity"], "host")?.into();
        }
    }
    Ok((
        json!({"format_version":3,"database":client.database,"http_url":client.url,
        "checkpoint_database":identifier(options.checkpoint_database.as_deref().unwrap_or(&client.database))?,
        "state_directory":directory.to_str().context("invalid source directory")?,"host":local_host,
        "native_target":format!("{}:{}/{}",parsed.host_str().context("missing native database host")?,parsed.port().unwrap_or(9000),client.database),
        "endpoint":options.endpoint,"accounts":selected,"start_block":options.start_block,"module":MODULE,
        "module_hash":string(module,"hash")?,"package_sha256":hex::encode(Sha256::digest(&package)),
        "network":"bsc","schema_version":1,"final_blocks_only":true}),
        package,
    ))
}

pub(crate) fn prepare_unlocked(
    client: &ClickHouse,
    options: &NativeOptions,
    directory: &Path,
) -> Result<Value> {
    // Reject changed local intent before invoking the native CLI. A changed
    // cohort/range/destination cannot become an external-command retry path.
    let previous_path = directory.join("run.json");
    if previous_path.try_exists()? {
        let previous: Value = serde_json::from_slice(&fs::read(previous_path)?)?;
        let previous = &previous["identity"];
        ensure!(
            previous["database"] == client.database
                && previous["http_url"] == client.url
                && previous["endpoint"] == options.endpoint
                && previous["start_block"] == options.start_block
                && previous["state_directory"]
                    == directory.to_str().context("invalid source directory")?
                && previous["accounts"]
                    == serde_json::to_value(canonical_accounts(&options.accounts)?)?
                && previous["checkpoint_database"]
                    == options
                        .checkpoint_database
                        .as_deref()
                        .unwrap_or(&client.database),
            "run identity changed; use a new isolated database and state directory"
        );
    }
    let (identity, package) = identity(client, options, directory)?;
    let record_path = directory.join("run.json");
    let admin = client.with_database("default")?;
    let query = params(json!({"db":client.database}))?;
    let exists = uint(
        &admin.one(
            "SELECT count() AS n FROM system.databases WHERE name={db:String}",
            &query,
        )?["n"],
    )? != 0;
    let mut record;
    if record_path.try_exists()? {
        record = serde_json::from_slice::<Value>(&fs::read(&record_path)?)?;
        ensure!(
            matches!(record["phase"].as_str(), Some("preparing" | "prepared")),
            "invalid native run phase"
        );
        ensure!(
            record["identity"] == identity,
            "run identity changed; use a new isolated database and state directory"
        );
        ensure!(
            record["phase"] != "prepared" || exists,
            "run database is missing; restore matching database and cursor metadata"
        );
    } else {
        if exists {
            ensure!(
                uint(
                    &client.one(
                        "SELECT count() AS n FROM system.tables WHERE database={db:String}",
                        &query
                    )?["n"]
                )? == 0,
                "database already contains tables without this run's local identity"
            );
        }
        for entry in fs::read_dir(directory)? {
            ensure!(
                entry?.file_name() == "run.lock",
                "state directory is not empty and has no run identity"
            );
        }
        record = json!({"run_id":new_id(),"identity":identity,"phase":"preparing"});
        atomic_json(&record_path, &record, false)?;
    }
    let frozen = directory.join("package.spkg");
    if frozen.try_exists()? {
        ensure!(
            hex::encode(Sha256::digest(fs::read(&frozen)?)) == string(&identity, "package_sha256")?,
            "frozen run package was changed"
        );
    } else {
        ensure!(
            record["phase"] != "prepared",
            "frozen run package is missing"
        );
        atomic_write(&frozen, &package, false)?;
    }
    admin.execute(
        &format!("CREATE DATABASE IF NOT EXISTS {}", client.database),
        &Default::default(),
    )?;
    let db = client.one(
        "SELECT toString(uuid) AS uuid FROM system.databases WHERE name={db:String}",
        &query,
    )?;
    ensure!(
        record
            .get("database_uuid")
            .is_none_or(|value| *value == db["uuid"]),
        "database was replaced; restore matching database and run metadata"
    );
    let owner_exists=uint(&client.one("SELECT count() AS n FROM system.tables WHERE database={db:String} AND name='_evm_state_run'",&query)?["n"])?!=0;
    if !owner_exists {
        ensure!(
            record["phase"] != "prepared",
            "database run identity is missing"
        );
        client.execute("CREATE TABLE _evm_state_run (run_id String, identity String) ENGINE=MergeTree ORDER BY run_id SETTINGS fsync_after_insert=1, fsync_part_directory=1",&Default::default())?;
        client.insert_values(
            "_evm_state_run",
            [json!({"run_id":record["run_id"],"identity":spaced_json(&identity)?})],
        )?;
    }
    let owner = client.one(
        "SELECT run_id,identity FROM _evm_state_run",
        &Default::default(),
    )?;
    ensure!(
        owner["run_id"] == record["run_id"]
            && serde_json::from_str::<Value>(string(&owner, "identity")?)? == identity,
        "database belongs to a different native run"
    );
    let meta = directory.join("meta");
    let metadata = meta.join(format!("{}_schema_hash.txt", client.database));
    if record["phase"] == "preparing" {
        fs::create_dir_all(&meta)?;
        native(
            &[
                "sink".into(),
                "clickhouse".into(),
                "setup".into(),
                frozen.to_str().context("invalid package path")?.into(),
                MODULE.into(),
                "--bytes-encoding".into(),
                "0xhex".into(),
                "--sink-info-folder".into(),
                meta.to_str().context("invalid metadata directory")?.into(),
            ],
            &options.dsn,
        )?;
        for table in ["state_blocks", "_blocks_"] {
            client.execute(&format!("ALTER TABLE {table} MODIFY SETTING fsync_after_insert=1, fsync_part_directory=1"),&Default::default())?;
        }
        fs::File::open(&metadata)?.sync_all()?;
        record["phase"] = json!("prepared");
        record["database_uuid"] = db["uuid"].clone();
        record["schema_hash"] = json!(fs::read_to_string(&metadata)?.trim());
        atomic_json(&record_path, &record, true)?;
    }
    ensure!(
        metadata.is_file()
            && fs::read_to_string(metadata)?.trim() == string(&record, "schema_hash")?,
        "native schema metadata is missing or changed"
    );
    Ok(record)
}

pub fn prepare(client: &ClickHouse, options: &NativeOptions) -> Result<Value> {
    let directory = resolve(&options.state_dir)?;
    capacity::check(client, &[directory.clone()], "native-prepare")?;
    let _lock = file_lock(&directory.join("run.lock"), true, false)?;
    prepare_unlocked(client, options, &directory)
}

pub fn recover_cursor(client: &ClickHouse, options: &NativeOptions) -> Result<Value> {
    let directory = resolve(&options.state_dir)?;
    capacity::check(client, &[directory.clone()], "native-recover")?;
    let _lock = file_lock(&directory.join("run.lock"), true, false)?;
    let run = prepare_unlocked(client, options, &directory)?;
    let progress = cursor::load_progress(client, &run, &directory)?;
    let cursor = directory.join("cursor.txt");
    if cursor.is_file() {
        atomic_write(
            &directory.join(format!("cursor-before-recovery-{}.txt", new_id())),
            &fs::read(&cursor)?,
            false,
        )?;
    }
    atomic_write(&cursor, string(&progress, "cursor")?.as_bytes(), true)?;
    Ok(json!({"run_id":run["run_id"],"recovered":true,"position":progress["position"]}))
}

pub fn command_args(
    options: &NativeOptions,
    ingest: &IngestOptions,
    record: &Value,
    directory: &Path,
) -> Result<Vec<String>> {
    ingest.validate(options.start_block)?;
    let text = |path: PathBuf| {
        path.to_str()
            .map(str::to_owned)
            .context("invalid native state path")
    };
    let accounts = record["identity"]["accounts"]
        .as_array()
        .context("missing native filter")?
        .iter()
        .map(|v| v.as_str().context("invalid native filter"))
        .collect::<Result<Vec<_>>>()?
        .join(",");
    let mut args = vec![
        "sink".into(),
        "clickhouse".into(),
        text(directory.join("package.spkg"))?,
        MODULE.into(),
        "-e".into(),
        options.endpoint.clone(),
        "-p".into(),
        format!("{MODULE}={accounts}"),
        "-s".into(),
        options.start_block.to_string(),
        "--final-blocks-only".into(),
        "--bytes-encoding".into(),
        "0xhex".into(),
        "--sink-info-folder".into(),
        text(directory.join("meta"))?,
        "--cursor-file-path".into(),
        text(directory.join("cursor.txt"))?,
        "--spool-dir".into(),
        text(directory.join("spool"))?,
        "--spool-max-size".into(),
        "1GiB".into(),
        "--spool-max-idle".into(),
        format!("{}ms", ingest.spool_max_idle_ms),
        "--max-retries".into(),
        ingest.max_retries.to_string(),
        "--decode-batch-size".into(),
        ingest.decode_batch_size.to_string(),
    ];
    if let Some(workers) = ingest.parallel_workers {
        args.extend([
            "--header".into(),
            format!("X-Substreams-Parallel-Workers:{workers}"),
        ]);
    }
    if let Some(address) = &ingest.prometheus_addr {
        args.extend(["--prometheus-addr".into(), address.clone()]);
    }
    if let Some(stop) = ingest.stop_block {
        args.extend(["-t".into(), stop.to_string()]);
    }
    Ok(args)
}

pub fn ingest(
    client: &ClickHouse,
    options: &NativeOptions,
    ingest: &IngestOptions,
) -> Result<Value> {
    ingest_bounded(client, options, ingest, None, None)
}

/// Optional qualification deadline/cancellation. The native child remains in
/// its supervisor's process group and is always reaped before returning.
pub fn ingest_bounded(
    client: &ClickHouse,
    options: &NativeOptions,
    ingest: &IngestOptions,
    deadline: Option<Instant>,
    cancelled: Option<&AtomicBool>,
) -> Result<Value> {
    let check_bound = || -> Result<()> {
        ensure!(
            deadline.is_none_or(|end| Instant::now() < end),
            "bounded native ingestion timed out; retain cursor and spool"
        );
        ensure!(
            !cancelled.is_some_and(|flag| flag.load(Ordering::Relaxed)),
            "native qualification cancelled; retain cursor and spool"
        );
        Ok(())
    };
    check_bound()?;
    ingest.validate(options.start_block)?;
    let directory = resolve(&options.state_dir)?;
    let lock = file_lock(&directory.join("run.lock"), true, false)?;
    capacity::check(client, &[directory.clone()], "native-start")?;
    let record = prepare_unlocked(client, options, &directory)?;
    check_bound()?;
    let cursor_path = directory.join("cursor.txt");
    let blocks = uint(
        &client.one(
            "SELECT count() AS n FROM state_blocks FINAL",
            &Default::default(),
        )?["n"],
    )?;
    ensure!(
        blocks == 0
            || cursor_path.is_file() && !fs::read_to_string(&cursor_path)?.trim().is_empty(),
        "native cursor is missing or empty for existing data; explicit recovery is required"
    );
    ensure!(
        blocks != 0 || !cursor_path.try_exists()?,
        "cursor exists without its block data; restore matching database and metadata"
    );
    let mut report = json!({"run_id":record["run_id"],"source":{"database":client.database,"accounts":record["identity"]["accounts"],
        "start_block":options.start_block,"module_hash":record["identity"]["module_hash"],"final_blocks_only":true}});
    if blocks != 0 {
        let progress = cursor::save_progress(
            client,
            &record,
            &directory,
            fs::read_to_string(&cursor_path)?.trim(),
        )?;
        if let Some(stop) = ingest.stop_block {
            let number = uint(&progress["position"]["block"]["number"])?;
            if number >= stop - 1 {
                ensure!(
                    number == stop - 1,
                    "requested stop precedes the already ingested cursor"
                );
                report["already_complete"] = json!(true);
                report["position"] = progress["position"].clone();
                return Ok(report);
            }
        }
    } else {
        ensure!(
            !directory.join("durable_progress.json").try_exists()?,
            "durable progress exists without its block data"
        );
    }
    let mut command = Command::new("substreams");
    command
        .args(command_args(options, ingest, &record, &directory)?)
        .env("SUBSTREAMS_SINK_DSN", &options.dsn);
    lock.inherit_in(&mut command);
    let mut child = process::ChildGuard(command.spawn()?);
    let stopped = Arc::new(AtomicBool::new(false));
    let signals = [signal_hook::consts::SIGINT, signal_hook::consts::SIGTERM]
        .into_iter()
        .map(|signal| signal_hook::flag::register(signal, stopped.clone()))
        .collect::<std::io::Result<Vec<_>>>()?;
    let run = (|| -> Result<()> {
        loop {
            check_bound()?;
            let status = child.wait_timeout(Duration::from_secs(1))?;
            check_bound()?;
            if let Some(status) = status {
                ensure!(
                    status.success(),
                    "native sink exited with status {}; retain its cursor and spool for recovery",
                    process::exit_code(status)
                );
                break;
            }
            ensure!(
                !stopped.load(Ordering::Relaxed),
                "native ingestion interrupted; retain its cursor and spool for recovery"
            );
            cursor::observe(client, &record, &directory)?;
            capacity::check(client, &[directory.clone()], "native-progress")?;
        }
        Ok(())
    })();
    for signal in signals {
        signal_hook::low_level::unregister(signal);
    }
    if child.try_wait()?.is_none() {
        // SAFETY: this PID belongs to our still-running native child.
        unsafe {
            libc::kill(child.id() as i32, libc::SIGTERM);
        }
        if child.wait_timeout(Duration::from_secs(15))?.is_none() {
            child.kill()?;
            child.wait()?;
        }
    }
    // A graceful stop may seal a spool and persist a newer valid cursor.
    let progress = cursor::observe(client, &record, &directory)?;
    run?;
    let progress = progress
        .context("native sink completed without a valid cursor; recover from durable progress")?;
    if let Some(stop) = ingest.stop_block {
        ensure!(
            uint(&progress["position"]["block"]["number"])? == stop - 1,
            "native sink stopped before its requested final block"
        );
    }
    fs::File::open(&cursor_path)?.sync_all()?;
    atomic_write(
        &directory.join("last_completed_cursor.txt"),
        string(&progress, "cursor")?.as_bytes(),
        true,
    )?;
    report["position"] = progress["position"].clone();
    Ok(report)
}
